src/core/types.ts

```diff
@@ -7,7 +7,13 @@ import type { MutationCache } from './mutationCache'
 import { Logger } from './logger'
 
 export type QueryKey = readonly unknown[]
-export type QueryFunctionData<T> = T extends undefined ? never : T
+export type QueryFunctionData<TReturn> = TReturn extends Promise<
+  infer TPromiseReturn
+>
+  ? Promise<QueryFunctionData<TPromiseReturn>>
+  : TReturn extends undefined
```

Comment:
Should we extend the conditional type here also checking for `void`?

Otherwise, query functions with explicit `void` return type will not fail

```typescript
useQuery(key, (): void => undefined) // no errors at the moment
useQuery(key, async (): Promise<void> => undefined) // no errors at the moment
```